packages/api/src/routers/external-api/v2/charts.ts

```diff
@@ -0,0 +1,351 @@
+import { ClickhouseClient } from '@hyperdx/common-utils/dist/clickhouse';
+import { getMetadata } from '@hyperdx/common-utils/dist/metadata';
+import {
+  ChartConfigWithOptDateRange,
+  DisplayType,
+} from '@hyperdx/common-utils/dist/types';
+import { SourceKind } from '@hyperdx/common-utils/dist/types';
+import opentelemetry, { SpanStatusCode } from '@opentelemetry/api';
+import express from 'express';
+import _ from 'lodash';
+import { z } from 'zod';
+import { validateRequest } from 'zod-express-middleware';
+
+import { Granularity } from '@/clickhouse';
+import { getConnectionById } from '@/controllers/connection';
+import { getSource } from '@/controllers/sources';
+import { getTeam } from '@/controllers/team';
+import { IConnection } from '@/models/connection';
+import { ISource } from '@/models/source';
+import { translateExternalSeriesToInternalSeries } from '@/utils/externalApi';
+import {
+  externalQueryChartSeriesSchema,
+  objectIdSchema,
+  timeChartSeriesSchema,
+} from '@/utils/zod';
+
+// Mapping from API's MetricsDataType enum (title-case) to common-utils format (lowercase)
+const normalizeMetricDataType = (metricType?: string): string | undefined => {
+  if (!metricType) return undefined;
+
+  // Convert first letter to lowercase to match common-utils expectations
+  return metricType.charAt(0).toLowerCase() + metricType.slice(1);
+};
+
+// Types
+type ExternalSeriesInput = z.infer<typeof externalQueryChartSeriesSchema>;
+
+const API_GRANULARITY_TO_INTERNAL: Record<string, Granularity> = {
+  '1s': '1 second' as Granularity, // For testing
+  '30s': Granularity.ThirtySecond,
+  '1m': Granularity.OneMinute,
+  '5m': Granularity.FiveMinute,
+  '10m': Granularity.TenMinute,
+  '15m': Granularity.FifteenMinute,
+  '30m': Granularity.ThirtyMinute,
+  '1h': Granularity.OneHour,
+  '2h': Granularity.TwoHour,
+  '6h': Granularity.SixHour,
+  '12h': Granularity.TwelveHour,
+  '1d': Granularity.OneDay,
+  '2d': Granularity.TwoDay,
+  '7d': Granularity.SevenDay,
+  '30d': Granularity.ThirtyDay,
+  auto: 'auto' as Granularity,
+};
+
+// Create schemas for different environments
+const productionGranularitySchema = z.enum([
+  'auto',
+  '30s',
+  '1m',
+  '5m',
+  '10m',
+  '15m',
+  '30m',
+  '1h',
+  '2h',
+  '6h',
+  '12h',
+  '1d',
+  '2d',
+  '7d',
+  '30d',
```

Comment:
is it possible to use the keys from the record above to produce this list? it seems funny that it's the same set of strings in both.

Reply:
I guess the only difference is that the test set allows 1 second of granularity.

Reply:
yeahhhh def will clean that up

Reply:
Addressed